parallelise contributor points calculations

matt-fidd opened the pull request:
Before:
```sh
% time node .github/scripts/count-points.mjs
0.95s user 0.10s system 0% cpu 2:21.68 total
```

After:
```sh
% time node .github/scripts/count-points.mjs
0.88s user 0.08s system 14% cpu 6.702 total
```

Down from 2m21 to 6s - a nice improvement!

Rate limit info here: https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits

> No more than 100 concurrent requests are allowed. This limit is shared across the REST API and GraphQL API.

90 should be safe here, obviously if you spam the command you'll hit the per-minute rate limit a bit faster now.

MikesGlitch reviewed (commented):
I had a look at the plimit source and I'm not sure if it'll handle nested promises like this. I think it requires you to wrap every single api call in a limit(() => call)

e.g.

  limit(() => octokit.pulls.listReviews(...)),
  limit(() => octokit.paginate(...)),

I haven't tested though, this might be good enough - not familiar with this lib

youngcw commented:
Im getting these errors when trying to run this


```
GET /repos/actualbudget/actual/pulls/5032/files?per_page=100 - 500 with id UNKNOWN in 10029ms
RequestError [HttpError]: Connect Timeout Error (attempted address: api.github.com:443, timeout: 10000ms)
    at fetchWrapper (file:///home/caleb/Desktop/actual/node_modules/@octokit/request/dist-bundle/index.js:76:26)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at runNextTicks (node:internal/process/task_queues:64:3)
    at process.processImmediate (node:internal/timers:454:9)
    at async Object.next (file:///home/caleb/Desktop/actual/node_modules/@octokit/plugin-paginate-rest/dist-bundle/index.js:45:28)
    at async Promise.all (index 1)
    at async file:///home/caleb/Desktop/actual/.github/scripts/count-points.mjs:145:42 {
  status: 500,
  request: {
    method: 'GET',
    url: 'https://api.github.com/repos/actualbudget/actual/pulls/5032/files?per_page=100',
    headers: {
      accept: 'application/vnd.github.v3+json',
      'user-agent': 'octokit-rest.js/22.0.0 octokit-core.js/7.0.2 Node.js/20.19.4 (linux; x64)',
      authorization: 'token [REDACTED]'
    },
    request: { hook: [Function: bound bound register] }
  },
  response: undefined,
  cause: TypeError: fetch failed
      at node:internal/deps/undici/undici:13510:13
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at runNextTicks (node:internal/process/task_queues:64:3)
      at process.processImmediate (node:internal/timers:454:9)
      at async fetchWrapper (file:///home/caleb/Desktop/actual/node_modules/@octokit/request/dist-bundle/index.js:50:21)
      at async Object.next (file:///home/caleb/Desktop/actual/node_modules/@octokit/plugin-paginate-rest/dist-bundle/index.js:45:28)
      at async Promise.all (index 1)
      at async file:///home/caleb/Desktop/actual/.github/scripts/count-points.mjs:145:42 {
    [cause]: ConnectTimeoutError: Connect Timeout Error (attempted address: api.github.com:443, timeout: 10000ms)
        at onConnectTimeout (node:internal/deps/undici/undici:2602:28)
        at Immediate._onImmediate (node:internal/deps/undici/undici:2583:11)
        at process.processImmediate (node:internal/timers:483:21) {
      code: 'UND_ERR_CONNECT_TIMEOUT'
    }
  }
}
GET /repos/actualbudget/actual/pulls/5319/reviews - 500 with id UNKNOWN in 10076ms
GET /repos/actualbudget/actual/pulls/5319/files?per_page=100 - 500 with id UNKNOWN in 10076ms
GET /repos/actualbudget/actual/pulls/5316/reviews - 500 with id UNKNOWN in 10075ms
GET /repos/actualbudget/actual/pulls/5316/files?per_page=100 - 500 with id UNKNOWN in 10075ms
GET /repos/actualbudget/actual/pulls/5309/reviews - 500 with id UNKNOWN in 10074ms
GET /repos/actualbudget/actual/pulls/5326/files?per_page=100 - 500 with id UNKNOWN in 10077ms
..... etc

matt-fidd commented:
> I had a look at the plimit source and I'm not sure if it'll handle nested promises like this. I think it requires you to wrap every single api call in a limit(() => call)
> 
> e.g.
> 
> limit(() => octokit.pulls.listReviews(...)), limit(() => octokit.paginate(...)),
> 
> I haven't tested though, this might be good enough - not familiar with this lib

I'm pretty sure it just limits the execution of any promise-returning function. So in this case, it would allow "n" parallel executions of single pull request/issue calculations, regardless of how many things happen inside that function.

That being said, with it running multiple requests within the limit, I've dropped the total allowed to 30, just to be safe with the rate limit.

> Im getting these errors when trying to run this. Only happens on the main repo stage

Hopefully this is just a poorly communicated rate limit. I've dropped the limit to 30 concurrent now, hopefully it should work.
I've managed to run it a lot in testing without hitting these errors somehow.

MikesGlitch reviewed (approved):
Nice
